File: 2015/D01/d01.js
/*
Advent Of Code 2015
Day 1: Not Quite Lisp part 1 & 2

https://adventofcode.com/2015/day/1
*/

const fs = require('fs');
const path = require('path');

const YEAR = '2015';
const DAY = '01';

function readInput(filename) {
  const readRawIntput = () => {
    const data = fs.readFileSync(path.join(`${YEAR}`, `D${DAY}`, filename));
    return data.toString();
  };

  const parseALine = (line) => {};

  try {
    return readRawIntput();
  } catch (err) {
    console.error(err);
  }
}

function part1(input) {
  function findFinalFloor(instructions) {
    let floor = 0;

    for (let i = 0; i < instructions.length; i++) {
      if (instructions[i] === '(') {
        floor += 1;
      } else if (instructions[i] === ')') {
        floor -= 1;
      }

      if (floor === -1) {
        // Return the current instruction index (adding 1 for 1-based counting)
        return i + 1;
      }
    }

    return floor;
  }
  const result = findFinalFloor(input);
  return result;
}

function part2(input) {
  const { maxThree } = input.reduce(
    ({ maxThree, sum }, n) => {
      if (Number.isInteger(n)) return { maxThree, sum: sum + n };

      if (sum > maxThree[2]) {
        maxThree.push(sum);
        maxThree.sort((a, b) => b - a);
        maxThree.pop();
      }
      return { maxThree, sum: 0 };
    },
    { maxThree: [0, 0, 0], sum: 0 }
  );
  return maxThree.reduce((sum, n) => sum + n, 0);
}

const input = readInput(`d${DAY}-input.txt`);
console.log(`Part 1: ${part1(input)}`);
console.log(`Part 2: ${part2(input)}`);

File: 2015/D01/d01.py
"""
Advent Of Code 2022
Day 1: Calorie Counting part 1 & 2

https://adventofcode.com/2022/day/1
"""
from time import perf_counter as pfc
import os
import functools

YEAR = "2022"
DAY = "01"

def readInput(filename):
  with open(os.path.join(f"{YEAR}", f"D{DAY}", filename)) as f:
    lines = f.readlines()
  input = (d.strip('\n') for d in lines)
  return tuple(input)

def part1(input):
  startTime = pfc()

  result = functools.reduce(
      lambda result, n: (result[0], result[1] + int(n))
      if n.isnumeric() else (max(result[0], result[1]), 0), input, (0, 0))

  print(f"Part 1 time: {pfc() - startTime:.4}")
  return result[0]

def part2(input):
  startTime = pfc()

  # check if n should be part of the array of max 3 values
  def maxThree(array, n):
    if n > array[2]:
      array[2] = n
      array.sort(reverse=True)
    return array

  result = functools.reduce(
      lambda result, n: (result[0], result[1] + int(n))
      if n.isnumeric() else (maxThree(result[0], result[1]), 0), input, ([0, 0, 0], 0))

  print(f"Part 2 time: {pfc() - startTime:.4}")
  return sum(result[0])

input = readInput(f"d{DAY}-input.txt")
print(f"Part 1 for puzzle input: {part1(input)}")
print(f"Part 2 for puzzle input: {part2(input)}")
